### PR TITLE
Udeng 2282 refresh awareness implement new scheme

### PR DESCRIPTION
**Pull request by sergio-costas**
Implements the new Refresh Awareness protocol.

DONE:
- notifications when a new update is available but blocked because it's running
- actions in the notifications of pending updates
- show progress window (but only with pulsing progress bar)
- show update complete message
- show app icon in progress window
- show true progress bar during updates
- use the visible name from the .desktop file whenever possible
- group multiple progress bars in a single window
- make "hide" button work
- change from progress mode to pulse mode after five seconds of inactivity
- tested inside a snap. Requires several patches to snapd:
    https://github.com/snapcore/snapd/pull/13931
    https://github.com/snapcore/snapd/pull/13933
    https://github.com/snapcore/snapd/pull/13936
- added failsafes to ensure that it works without those patches (although not so "cute")
- added an icon for the task bar (it shows a generic icon when a window is shown)

TO DO:
- add an icon for upper-left notifications corner (it shows a generic icon)
- add progress bar in the icon on the dock

### CODE REVIEW

**Comment by sergio-costas** [comment]
@robert-ancell Should I add support for FreeDesktop.org notifications, or it should be enough with the gnome ones?

**Comment by robert-ancell** [comment]
> @robert-ancell Should I add support for FreeDesktop.org notifications, or it should be enough with the gnome ones?

Isn't Gio.Notification doing this for you? You don't seem to be making any explicit fdo/gnome notification calls that I can see.

**Comment by sergio-costas** [comment]
I don't know if Gio.Notification does abstract up to that point... I'll try to do some tests.

**Comment by sergio-costas** [comment]
You are right: it is supported. So nothing to do there, then. Thanks!

**Comment by sergio-costas** [comment]
@robert-ancell This is ready, with the exception of an icon for the application itself (because the notifications show the application icon too, in the top-left corner). Which one should I use? The "snapcraft" one? The "snap store" one?

**Comment by sergio-costas** [comment]
Also, it is "compatible" with the current snapd version. This is: it won't be as "cute" as it will be when all the patches have landed in snapd, but it works. It doesn't show the application icon or the "visible name", and the "application is ready" notification is shown twice, but it works.

**Comment by sergio-costas** [comment]
Also, the `desktop-launch` interface is there just for when those patches land, but... I still don't know if that will be the definitive interface for reading the .desktop files, or will have to use a different one. If you want to launch this "as-is", maybe it would be a good idea to remove it until that patch lands... What do you think?

**Comment by robert-ancell** [comment]
> Also, the `desktop-launch` interface is there just for when those patches land, but... I still don't know if that will be the definitive interface for reading the .desktop files, or will have to use a different one. If you want to launch this "as-is", maybe it would be a good idea to remove it until that patch lands... What do you think?

Yes, if it can be removed now do it an make a second PR to add it.

File: src/sdi-snap.c
Comment (robert-ancell) [inline src/sdi-snap.c:72]: spelling hiden -> hidden

File: src/sdi-snap.c
Comment (robert-ancell) [inline src/sdi-snap.c:99]: Blank line here inconsistent (other functions don't have it). Same below.

File: src/sdi-snap.c
Comment (robert-ancell) [inline src/sdi-snap.c:159]: Could use g_clear_object.

File: src/sdi-snap.h
Comment (robert-ancell) [inline src/sdi-snap.h:25]: Nit: I have not bothered to make these defines - you can just use the _get_type() function directly (as you have).

File: src/sdi-theme-monitor.h
Comment (robert-ancell) [inline src/sdi-theme-monitor.h:25]: Same nit about unused define.

File: src/sdi-snap.c
Comment (robert-ancell) [inline src/sdi-snap.c:56]: Use g_clear_object here and then you can simplify the if below to just `if (dialog != NULL) ...`

File: src/sdi-snap.c
Comment (robert-ancell) [inline src/sdi-snap.c:42]: Does this need to be a ref_sink here - dialog is only set to a full reference in sdi_snap_set_dialog?

File: src/sdi-snap.c
Comment (robert-ancell) [inline src/sdi-snap.c:25]: Would be nice to have some comments on these fields.

File: src/sdi-refresh-monitor.c
Comment (robert-ancell) [inline src/sdi-refresh-monitor.c:58]: Can just merge this into the above type, i.e. `typedef struct { ... } SnapRefreshData;`

File: src/sdi-refresh-monitor.c
Comment (robert-ancell) [inline src/sdi-refresh-monitor.c:60]: static

File: src/sdi-refresh-monitor.c
Comment (robert-ancell) [inline src/sdi-refresh-monitor.c:76]: Remove empty line

File: src/sdi-refresh-monitor.c
Comment (robert-ancell) [inline src/sdi-refresh-monitor.c:392]: Expand comment to indicate why this is not required to be implemented now and when it would in the future.

**Comment by robert-ancell** [comment]
Not a full review, we can go over again next week.

File: src/sdi-snap.c
Comment (sergio-costas) [reply src/sdi-snap.c]: Ops... fixed everywhere.

File: src/sdi-snap.c
Comment (sergio-costas) [reply src/sdi-snap.c:99]: Both fixed.

File: src/sdi-snap.c
Comment (sergio-costas) [reply src/sdi-snap.c:159]: Fixed.

File: src/sdi-snap.h
Comment (sergio-costas) [reply src/sdi-snap.h:25]: Yes, but for consistency with GObject I think that it's a good idea to use them.

Comment (sergio-costas) [reply src/sdi-snap.h:25]: (mainly for the cases where properties are used).

File: src/sdi-snap.c
Comment (sergio-costas) [reply src/sdi-snap.c:42]: I had a problem with the unsinken references (the dialogues got freed before than they should), so I decided to use ref_sink everywhere just to avoid problems... But maybe it's better to just return a sinked reference in the constructor.

File: src/sdi-snap.c
Comment (sergio-costas) [reply src/sdi-snap.c:25]: Good point. Done.

File: src/sdi-refresh-monitor.c
Comment (sergio-costas) [reply src/sdi-refresh-monitor.c:58]: Done.

File: src/sdi-refresh-monitor.c
Comment (sergio-costas) [reply src/sdi-refresh-monitor.c:76]: Fixed.

File: src/sdi-refresh-monitor.c
Comment (sergio-costas) [reply src/sdi-refresh-monitor.c]: Done.

File: src/sdi-refresh-monitor.c
Comment (sergio-costas) [reply src/sdi-refresh-monitor.c]: Done.

File: src/sdi-refresh-monitor.c
Comment (sergio-costas) [reply src/sdi-refresh-monitor.c]: Done.

File: src/sdi-refresh-monitor.c
Comment (sergio-costas) [reply src/sdi-refresh-monitor.c]: Good point. Done.

File: src/sdi-snap.c
Comment (sergio-costas) [reply src/sdi-snap.c:26]: Added. Check them to see if they are fine, please.

File: src/main.c
Comment (robert-ancell) [inline src/main.c:142]: (not caused by this PR). Just noting the GError is set to NULL - it should be caught and written into the g_message below.

File: src/sdi-helpers.c
Comment (robert-ancell) [inline src/sdi-helpers.c:26]: Nit: You could have less indentation by returning early from the function is `apps->len < 1`.

File: src/sdi-notify.c
Comment (robert-ancell) [inline src/sdi-notify.c:69]: Nit: ID should be lowercase.

File: src/sdi-notify.c
Comment (robert-ancell) [inline src/sdi-notify.c:83]: Nit: these should be just above where they are set.

File: src/sdi-notify.c
Comment (robert-ancell) [inline src/sdi-notify.c:96]: Nit, use defines for these values

File: src/sdi-notify.c
Comment (robert-ancell) [inline src/sdi-notify.c:191]: Left over debugging? Perhaps change to a g_warning/g_debug if you want this to remain.

File: src/sdi-notify.c
Comment (robert-ancell) [inline src/sdi-notify.c:195]: Nit: empty line.

File: src/sdi-refresh-monitor.c
Comment (robert-ancell) [inline src/sdi-refresh-monitor.c:43]: Nit: ID -> id

File: src/sdi-refresh-monitor.c
Comment (robert-ancell) [inline src/sdi-refresh-monitor.c:232]: Nit: empty line

File: src/sdi-refresh-monitor.c
Comment (robert-ancell) [inline src/sdi-refresh-monitor.c:333]: GPtrArray.len is guint

**Review by robert-ancell — APPROVED** [review]
Some minor notes, I'll leave it to you to decide if they're worth changing. (feel free to merge when you're ready).

The change is large, so I'm not 100% confident my review is exhaustive - hopefully no major issues in here!

File: src/sdi-helpers.c
Comment (sergio-costas) [reply src/sdi-helpers.c:26]: Good point. Done.

File: src/main.c
Comment (sergio-costas) [reply src/main.c:142]: You are right. Done.

File: src/sdi-notify.c
Comment (sergio-costas) [reply src/sdi-notify.c:69]: Fixed.

File: src/sdi-notify.c
Comment (sergio-costas) [reply src/sdi-notify.c:83]: Done.

File: src/sdi-notify.c
Comment (sergio-costas) [reply src/sdi-notify.c:96]: Fixed.

File: src/sdi-notify.c
Comment (sergio-costas) [reply src/sdi-notify.c:191]: Yes, a left over O:-) Removing.

File: src/sdi-notify.c
Comment (sergio-costas) [reply src/sdi-notify.c:195]: Removed.

File: src/sdi-refresh-monitor.c
Comment (sergio-costas) [reply src/sdi-refresh-monitor.c:43]: ops... Fixed.

File: src/sdi-refresh-monitor.c
Comment (sergio-costas) [reply src/sdi-refresh-monitor.c:232]: Removed too.

File: src/sdi-refresh-monitor.c
Comment (sergio-costas) [reply src/sdi-refresh-monitor.c:333]: Ops... fixed there an in other place.

**Comment by sergio-costas** [comment]
Done the changes. But I can't merge it, I don't have permissions.

**Comment by sergio-costas** [comment]
Anyway, there are two things remaining.

**Comment by robert-ancell** [comment]
Ok, ping me when you're ready to merge.

**Comment by sergio-costas** [comment]
@robert-ancell I think that it's better to merge this one, and do those other changes in different PRs, so please, merge this one when you have some spare time.

**Review by robert-ancell — APPROVED** [review]

**Pull request merged**
